### PR TITLE
Issue 23

### PR DESCRIPTION
**Pull request by achille-roussel**
This PR adds a test to verify the bug reported in #23

### CODE REVIEW

File: json/json_test.go [hunk 1]
```diff
@@ -1378,3 +1378,45 @@ func TestGithubIssue18(t *testing.T) {
...
 	}
 }
+
+func TestGithubIssue23(t *testing.T) {
+	t.Run("", func(t *testing.T) {
+		type d struct {
+			S map[string]string
+		}
+
+		b, _ := json.Marshal(map[string]d{
```
Comment (Gobd) [inline json/json_test.go]: This is using "encoding/json" & maybe other things are in these tests too that shouldn't be?

File: json/json_test.go [hunk 2]
```diff
@@ -1378,3 +1378,45 @@ func TestGithubIssue18(t *testing.T) {
...
+	t.Run("", func(t *testing.T) {
+		type testInner struct {
+			InnerMap map[string]string `json:"inner_map"`
+		}
+
+		type testOuter struct {
+			OuterMap map[string]testInner `json:"outer_map"`
+		}
+
+		b, _ := json.Marshal(testOuter{
```
Comment (ploxiln) [inline json/json_test.go]: typo: should be `Marshal` instead of `json.Marshal` (to test this package)

Comment (ploxiln) [reply json/json_test.go]: ... actually looks like it should use `marshal` to use the test harness switch between stdlib and this implementation

File: json/json_test.go (hunk 1 above)
Comment (ploxiln) [reply json/json_test.go]: Searching the file for "json." it looks like most uses are just as a reference/comparison ... maybe just `TestGithubIssue15` also has this mixup

Comment (ploxiln) [reply json/json_test.go]: ... and `TestDecodeStructFieldCaseInsensitive` ... that's probably it

Comment (achille-roussel) [reply json/json_test.go]: Thanks! I fixed all the misuses of the stdlib json package 👍

File: json/json_test.go [hunk 3]
```diff
@@ -336,7 +336,7 @@ func TestCodec(t *testing.T) {
 		t.Run(testName(v1), func(t *testing.T) {
 			v2 := newValue(v1)
 
-			a, err := json.MarshalIndent(v1, "", "\t")
+			a, err := MarshalIndent(v1, "", "\t")
```
Comment (ploxiln) [inline json/json_test.go]: I think this was intended to use the stdlib json package. Below it compares `a` (from the stdlib json) to `b` (from NewEncoder() from this package)

File: json/json_test.go [hunk 4]
```diff
@@ -499,7 +499,7 @@ func BenchmarkUnmarshal(b *testing.B) {
 				x = duration(d)
 			}
 
-			j, _ := json.Marshal(x)
+			j, _ := Marshal(x)
```
Comment (ploxiln) [inline json/json_test.go]: I think this intended to use the stdlib json package to marshal, and then this package to unmarshal. Below `unmarshal` is (by default) a pointer to this package's Unmarshal.

File: json/json_test.go (hunk 3 above)
Comment (achille-roussel) [reply json/json_test.go]: Possibly, this test hasn't been touched in a while, I think it predates adding the `MarshalIndent` function to this package. We still rely on `json.Ident` form stdlib so the behavior didn't change.

I'll revert if you think it's serves a better purpose the way it was 👍

Comment (ploxiln) [reply json/json_test.go]: I see, that also makes sense: comparing `MashalIndent` to `Encoder.SetIndent`. I'm no expert here, just trying to help :)

**Comment by achille-roussel** [comment]
@ploxiln @Gobd thanks for helping with the reviews!

This PR should have a fix for the bug your reported, if you're able to verify whether it addresses the issue or not let me know!

File: json/codec.go [hunk 5]
```diff
@@ -396,8 +396,7 @@ func constructMapCodec(t reflect.Type, seen map[reflect.Type]*structType) codec
 		}
 	}
 
-	switch v.Kind() {
-	case reflect.Ptr, reflect.Map:
+	if inlined(v) {
```
Comment (achille-roussel) [inline json/codec.go]: This was the issue.

**Comment by ploxiln** [comment]
much thanks for the rather quick fix!
cc @KoreyBarton

File: json/json_test.go [hunk 6]
```diff
@@ -1378,3 +1378,44 @@ func TestGithubIssue18(t *testing.T) {
...
+		}
+
+		type testOuter struct {
+			OuterMap map[string]testInner `json:"outer_map"`
+		}
+
+		b, _ := Marshal(testOuter{
+			OuterMap: map[string]testInner{
+				"outer": {
+					InnerMap: map[string]string{"inner": "value"},
```
Comment (jnjackins) [inline json/json_test.go]: Should we also add a test for a triple nested map, while we're here?

Comment (achille-roussel) [reply json/json_test.go]: Sure I can add this 👍

**Pull request merged**